File: lib/collections/Stack.cc
#include <cstddef>
#include <cstring>
#include <iostream>
#include <string>

template <typename E> class Stack {
private:
  int STACK_SIZE_;
  int pos_ = -1;

public:
  bool isUnbound;

private:
  E *stack_;

public:
  Stack(int initialSize, bool isUnbound)
      : STACK_SIZE_(initialSize), isUnbound(isUnbound) {
    stack_ = new E[STACK_SIZE_];
    std::memset(stack_, 0, sizeof(E) * STACK_SIZE_);
  };

  ~Stack() { delete[] stack_; }

  void ensureCapacity() {
    if (isUnbound) {
      if (pos_ == STACK_SIZE_ - 1) {
        E *newStack = new E[STACK_SIZE_ << 1];
        std::memcpy(newStack, stack_, STACK_SIZE_ * sizeof(E));
        delete[] stack_;

        stack_ = newStack;
        STACK_SIZE_ <<= 1;
      }
    }
  }

  E pop() {
    if (pos_ == -1)
      throw pos_;
    return stack_[pos_--];
  }

  void push(E element) {
    if (!isUnbound && pos_ >= STACK_SIZE_ - 1)
      throw STACK_SIZE_;
    ensureCapacity();
    stack_[++pos_] = element;
  }

  bool isEmpty() { return pos_ == -1; }
  size_t size() { return pos_ + 1; }
  void clear() {
    std::memset(stack_, 0, sizeof(E) * STACK_SIZE_);
    pos_ = -1;
  }

  E peek() {
    if (pos_ == -1)
      throw pos_;
    return stack_[pos_];
  }

  std::string to_string() {
    std::string res = "[";
    for (int i = 0; i <= pos_; i++)
      res += std::to_string(stack_[i]) + (i == pos_ ? "" : ", ");
    return res + "]";
  }
};

File: Library/java/src/collections/Stack.java
package collections;

import java.util.Arrays;
import java.util.EmptyStackException;

public class Stack<E> {
  public final boolean isUnbound;
  private int STACK_SIZE;
  private int pos = -1;
  private E[] stack;

  @SuppressWarnings("unchecked")
  public Stack(int initialSize, boolean isUnbound) {
    if (initialSize <= -1)
      throw new IllegalArgumentException();
    this.STACK_SIZE = initialSize;
    this.stack = (E[]) new Object[this.STACK_SIZE];
    this.isUnbound = isUnbound;
  }

  public void push(E element) {
    if (!this.isUnbound && this.pos >= this.STACK_SIZE - 1)
      throw new IllegalStateException();
    this.ensureCapacity();
    this.stack[++this.pos] = element;
  }

  public E pop() {
    if (this.pos == -1)
      throw new EmptyStackException();
    return this.stack[this.pos--];
  }

  public boolean isEmpty() {
    return this.pos == -1;
  }

  public int size() {
    return this.pos + 1;
  }

  public void clear() {
    Arrays.fill(this.stack, null);
    this.pos = -1;
  }

  public E peek() {
    if (this.pos == -1)
      throw new EmptyStackException();
    return this.stack[this.pos];
  }

  @Override
  public String toString() {
    final StringBuilder sb = new StringBuilder("[");
    for (int i = 0; i <= this.pos; i++)
      sb.append(String.valueOf(this.stack[i])).append(i == this.pos ? "" : ", ");
    return sb.append("]").toString();
  }

  @SuppressWarnings("unchecked")
  private void ensureCapacity() {
    if (this.isUnbound) {
      if (pos == this.STACK_SIZE - 1) {
        E[] newStack = (E[]) new Object[this.STACK_SIZE << 1];
        System.arraycopy(this.stack, 0, newStack, 0, this.STACK_SIZE);
        this.stack = newStack;
        this.STACK_SIZE <<= 1;
      }
    }
  }
}
